executor/point_get.go

```diff
@@ -236,13 +248,120 @@ func (e *PointGetExecutor) getAndLock(ctx context.Context, key kv.Key) (val []by
 	if err != nil {
 		return nil, err
 	}
-	val, err = e.get(ctx, key)
-	if err != nil && !kv.ErrNotExist.Equal(err) {
-		return nil, err
+	if !e.ctx.GetSessionVars().IsPessimisticReadConsistency() {
+		val, err = e.get(ctx, key)
+		if err != nil && !kv.ErrNotExist.Equal(err) {
+			return nil, err
+		}
 	}
 	return val, nil
 }
 
+func (e *PointGetExecutor) checkUniqueIndex(key []byte, chk *chunk.Chunk) error {
+	indexValues := make([]types.Datum, 0, 4)
+	for _, idxCol := range e.idxInfo.Columns {
+		colIdx := -1
+		colID := e.tblInfo.Columns[idxCol.Offset].ID
+		var schemaCol *expression.Column
+		for idx, colInSchema := range e.Schema().Columns {
+			if colInSchema.ID == colID {
+				colIdx = idx
+				schemaCol = colInSchema
+				break
+			}
+		}
+		if schemaCol != nil {
+			val := chk.GetRow(0).GetDatum(colIdx, schemaCol.RetType)
+			indexValues = append(indexValues, val)
+		}
+	}
+	conflictErr := kv.ErrWriteConflict.FastGenByArgs(e.startTS, 0, 0, key)
+	if len(indexValues) != len(e.idxVals) {
+		return conflictErr
+	}
+	for i, readlVal := range indexValues {
+		cmpRes, err := readlVal.CompareDatum(e.ctx.GetSessionVars().StmtCtx, &e.idxVals[i])
+		if err != nil {
+			return conflictErr
+		}
+		if cmpRes != 0 {
+			return conflictErr
+		}
+	}
+	return nil
+}
+
+// forceLockKey will try to lock the key even if the key value commitTS greater than forUpdateTS, and the
+// key will be locked and the value will be returned, some checks are needed to ensure the index and row
+// key consistency
+func (e *PointGetExecutor) forceLockKey(ctx context.Context, key []byte, chk *chunk.Chunk) error {
```

Comment:
should we embed it into `lockKeyIfNeeded`?

Reply:
Then the code would have too much `if` `else` branches

Reply:
@jackysp @coocood PTAL

Reply:
got